File: frontend/src/data/erpnext.js
// Live ERPNext DocType list for the Map Step `erpnext_doctype` picker. Fetched
// once (cached) and filtered client-side by the Combobox. Backing API:
// flowlane.api.erpnext.get_doctypes. Options are [{label, value}].

import { createResource } from 'frappe-ui'

export const doctypes = createResource({
  url: 'flowlane.api.erpnext.get_doctypes',
  cache: 'flowlane-doctypes',
})

export function doctypeOptions() {
  return doctypes.data || []
}

File: flowlane/api/erpnext.py
# Copyright (c) 2026, Onebook and contributors
# For license information, please see license.txt
"""ERPNext metadata lookups for the Table tab dropdowns.

The Map Step ``erpnext_doctype`` cell is a live DocType picker (no master — the
DocType list is metadata). ``get_doctypes`` returns the selectable, real DocTypes
as ``[{label, value}]`` for a frappe-ui Combobox; an optional ``module`` narrows
the list to one ERPNext module.
"""

import frappe


@frappe.whitelist()
def get_doctypes(module: str | None = None) -> list[dict]:
	"""Selectable DocTypes (no child tables, single, or virtual doctypes)."""
	filters = {"istable": 0, "issingle": 0, "custom": 0}
	if module and frappe.db.exists("Module Def", module):
		filters["module"] = module
	names = frappe.get_all("DocType", filters=filters, pluck="name", order_by="name asc")
	return [{"label": name, "value": name} for name in names]
